src/lib.js

```diff
@@ -88,8 +86,42 @@ export const query = (db, { select, where }) => {
     }
   }
 
-  const matches = evaluate(db, {
-    And: clauses.sort(byClause),
+
+  /**
+   * Ferret out all Cases, inclusive of those nested within Not and Rule clauses
+   * to ensure retreival of all related facts.
+   */
+
+  const cases = []
+
+  /** @type {Array<API.Clause>} */
+  const clauseStack = [...clauses]
+
+  while (clauseStack.length) {
+    const clause = clauseStack[0]
+    clauseStack.shift()
+    if (clause.Case) {
+      cases.push(clause.Case)
+    } else if (clause.And) {
+      clauseStack.unshift(...clause.And)
+    } else if (clause.Or) {
+      clauseStack.unshift(...clause.Or)
+    } else if (clause.Not) {
+      clauseStack.unshift(clause.Not)
+    } else if (clause.Rule) {
+      clauseStack.unshift(clause.Rule.rule.where)
+    }
+  }
+
+  // Collect all pertinent facts
+  const facts = await Promise.all(cases.map(c => iterateFacts(db, c)))
+
+  // Aggregate and de-dupe
+  const factMap = new Map()
+  facts.forEach(factArray => [...factArray].forEach(fact => factMap.set(encode(fact).toString(), fact)))
```

Comment:
Would be nice if we encapsulated deduping logic into an iterator to avoid having to hold all of the records in memory as number of records may make it prohibitive. Although I have to admit it is tricky here as same facts are scanned through by each subquery.

Also could you please use `Fact.link` function instead of `encode` here 

https://github.com/Gozala/datalogia/blob/32e7f5981ea33f719b1c20d22faf12189fe54379/src/fact.js#L73-L77